doc/reference/optimization.rst

```diff
@@ -0,0 +1,91 @@
+Optimization techniques for Plutus scripts
+==========================================
+
+Identifying problem areas
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In order to identify which parts of the script are responsible for significant resource consumption, you can use the :ref:`profiling support <profiling_scripts>`.
+
+Using strict let-bindings to avoid recomputation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Let-bindings in Haskell are translated to strict let-bindings in Plutus IR, unless they look like they might do computation, in which case they are translated to non-strict let-bindings.
+This is to avoid triggering effects (e.g. errors) at unexpected times.
+However, non-strict let-bindings are less efficient, so you may wish to explicitly mark let-bindings as strict in Haskell to avoid this.
+
+.. code-block:: haskell
+
+    -- This may be compiled non-strictly, which could result
+    -- in it being evaluated multiple times. However, it will
+    -- not be evaluated if it is not used.
+    let x = y + z
+    in if b then x + x else 1
+
+    -- This will be compiled strictly, but this will mean it
+    -- is evaluated even if it is not used.
+    let !x = y + z
+    in if b then x + x else 1
+
+Specializing higher-order functions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The use of higher-order functions is a common technique to facilitate code reuse.
+Higher-order functions are widely used in the Plutus libraries but can be less efficient than specialized versions.
+
+For instance, the Plutus function ``findOwnInput`` makes use of the higher-order function ``find`` to search for the current script input.
+
+.. code-block:: haskell
+
+    findOwnInput :: ScriptContext -> Maybe TxInInfo
+    findOwnInput ScriptContext{scriptContextTxInfo=TxInfo{txInfoInputs},
+                 scriptContextPurpose=Spending txOutRef} =
+        find (\TxInInfo{txInInfoOutRef} -> txInInfoOutRef == txOutRef) txInfoInputs
+    findOwnInput _ = Nothing
+
+This can be rewritten with a recursive function specialized to the specific check in question.
```

Comment:
Can rewrite rules help for not having to modify `findOwnInput`? https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/rewrite_rules.html?highlight=specialization%20rules#specialisation

Reply:
I don't think GHC will do *value* specialization in this way. But anyway, we mostly turn off GHC's optimizer, sadly :(